File: GenericSort.hs
import Data.Function
import Data.List

data Person = Person String Int deriving (Show, Eq)

name :: Person -> String
name   (Person n a) = n

age :: Person -> Int
age    (Person n a) = a 

instance Ord Person where
  p <= q = name p <= name q

numbers = [645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25]
people = [
  Person "Hal" 20,
  Person "Susann" 31,
  Person "Dwight" 19,
  Person "Kassandra" 21,
  Person "Lawrence" 25,
  Person "Cindy" 22,
  Person "Cory" 27,
  Person "Mac" 19,
  Person "Romana" 27,
  Person "Doretha" 32,
  Person "Danna" 20,
  Person "Zara" 23,
  Person "Rosalyn" 26,
  Person "Risa" 24,
  Person "Benny" 28,
  Person "Juan" 33,
  Person "Natalie" 25]

sortedNumbers = sortBy (\a b-> compare a b) numbers
nameSort = sortBy (compare `on` name) people
ageSort = sortBy ((flip compare `on` age) <> (compare `on` name)) people

main :: IO ()
main = do
  putStrLn "Sorted numberical values:"
  print sortedNumbers
  putStrLn "People sorted alphabetically by name:"
  print nameSort
  putStrLn "People sorted first by descending age and then alphabetical names:"
  print ageSort

File: GenericSort.py
# Class object type for People
class People:
  
  def __init__(self, personName, personAge ):
    self.name = personName
    self.age = personAge

  def __str__(self):
    return self.name + ", " + str(self.age)

  # sorts by age in descending order and compares name if age is equal as default type comarison
  def __lt__(self, person):
    # If the age is less return false, else true
    if (self.age < person.age):
      return False
    # If they are equal, switch to comparing the name
    elif (self.age == person.age):
      # If the name is less return true, else false
      if (self.name < person.name):
        return True
      return False
    return True


# Comparison function that only compares the name of an object as long as it has the name attribute
def compareNameOnly(object):
  return object.name


def main():
  # Array of people
  personArray = [People("Hal", 20), People("Susann", 31), People("Dwight", 19), People("Kassandra", 21), People("Lawrence", 25), People("Cindy", 22), People("Cory", 27), People("Mac", 19), People("Romana", 27), People("Doretha", 32), People("Danna", 20), People("Zara", 23), People("Rosalyn", 26), People("Risa", 24), People("Benny", 28), People("Juan", 33), People("Natalie", 25)]

  # Array of numbers
  numberArray = [645.41, 37.59, 76.41, 5.31, -34.23, 1.11, 1.10, 23.46, 635.47, -876.32, 467.83, 62.25]
  
  print("--------------------------------------------")
  print("Sorting age in descending order, and if equal then by name in lexicographical order")
  print("-------------")
  print("Before:")
  print("-------------")
  for i in range(0, len(personArray)):
      print(personArray[i], end=";\n")
  print("-------------")
  print("After:")
  print("-------------")
  # Uses generic sort
  personArray = sorted(personArray)
  for i in range(0, len(personArray)):
    print(personArray[i], end=";\n")
  print("--------------------------------------------")
  
  print("\n\n\n")
  
  print("--------------------------------------------") 
  print("Sorting name in lexicographical order")
  print("-------------")
  print("Before:")
  print("-------------")
  for i in range(0, len(personArray)):
      print(personArray[i], end=";\n")
  print("-------------")
  print("After:")
  print("-------------")
  # Uses generic sort, and compares the array
  # by using the compareNameOnly function that's passed in
  # as the key
  personArray = sorted(personArray, key=compareNameOnly)
  for i in range(0, len(personArray)):
    print(personArray[i], end=";\n")
  print("--------------------------------------------")
  
  print("\n\n\n")
  
  print("--------------------------------------------")
  print("Sorting numbers ascending by numerical value")
  print("-------------")
  print("Before:")
  print("-------------")
  for i in range(0, len(numberArray)):
      print(numberArray[i], end="\n")
  print("-------------")
  print("After:")
  print("-------------")
  # Uses generic sort
  numberArray = sorted(numberArray)
  for i in range(0, len(numberArray)):
    print(numberArray[i], end="\n")
  print("--------------------------------------------")

main()
